File: test/devices/garmin/garmin_settings_link_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:openvitals/devices/garmin/garmin_ble_transport.dart';
import 'package:openvitals/devices/garmin/garmin_session.dart';
import 'package:openvitals/devices/garmin/garmin_settings_link.dart';

/// Regression for "stop waiting on a link that has gone" (7b7858978).
///
/// Tapping an alarm hung on "Reading from the watch": the STATE request got no
/// reply, the link dropped ten seconds later, and the pending request sat out
/// its full thirty-second timeout waiting for an answer that could no longer
/// come. A dropped or closed link must resolve everything in flight at once.
void main() {
  GarminSettingsLink link() => GarminSettingsLink.forTest(
        GarminBleTransport(address: 'AA:BB:CC:DD:EE:FF'),
        GarminSession(
          // The watch never answers — every request would wait out its timeout.
          send: (_) async {},
          bluetoothName: 'Pixel 6 Pro',
          manufacturer: 'Google',
          model: 'raven',
          syncFiles: false,
        ),
      );

  test('closing the link resolves an in-flight screen read at once', () async {
    final l = link();
    // 65600 is the alarm screen id from the original hang.
    final pending = l.screen(65600);
    await pumpEventQueue();

    await l.close();

    // Well under GarminSettingsService.replyTimeout (30s): on regression this
    // throws TimeoutException instead of letting the test wait the hang out.
    final screen = await pending.timeout(const Duration(seconds: 2));
    expect(screen, isNull);
  });

  test('a request on an already-closed link answers null immediately', () async {
    final l = link();
    await l.close();

    expect(l.isOpen, isFalse);
    final screen = await l.screen(65600).timeout(const Duration(seconds: 2));
    expect(screen, isNull);
  });
}

File: app/src/test/kotlin/tech/mmarca/openvitals/devices/garmin/GarminSettingsLinkTest.kt
package tech.mmarca.openvitals.devices.garmin

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.async
import kotlinx.coroutines.test.runCurrent
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Regression for "stop waiting on a link that has gone" (7b7858978): a dropped link must
 * resolve everything in flight at once. These assert the read completes without advancing the clock.
 */
@OptIn(ExperimentalCoroutinesApi::class)
class GarminSettingsLinkTest {

    private fun link(scope: CoroutineScope): GarminSettingsLink =
        GarminSettingsLink.forTest(
            scope,
            GarminSession(
                scope = scope,
                // The watch never answers.
                send = {},
                bluetoothName = "Pixel 6 Pro",
                manufacturer = "Google",
                model = "raven",
                syncFiles = false,
            ),
        )

    @Test
    fun `closing the link resolves an in-flight screen read at once`() = runTest {
        val l = link(backgroundScope)
        // 65600 is the alarm screen id from the original hang.
        val pending = async { l.screen(65600) }
        runCurrent()

        l.close()
        runCurrent()

        // Resolved with no time advanced. On regression this waits out the 30-second timeout.
        assertTrue(pending.isCompleted)
        assertNull(pending.await())
    }

    @Test
    fun `a request on an already-closed link answers null immediately`() = runTest {
        val l = link(backgroundScope)
        l.close()
        runCurrent()

        assertFalse(l.isOpen)
        val pending = async { l.screen(65600) }
        runCurrent()
        assertTrue(pending.isCompleted)
        assertNull(pending.await())
    }
}
